text/0000-dynamic-import-hints.md

```diff
@@ -0,0 +1,257 @@
+
+#### Examples
+Valid
+```js
+// Single hint
+import("c-bar-mobile"/* "@salesforce/client/formFactor": "SMALL"*/);
+import("c-bar-mobile"/* "KEY.SUBKEY": "VALUE"*/);
+import("c-bar-mobile"/* "KEY/SUBKEY1": "VALUE"*/);
+import("c-bar-mobile"/* "KEY": "true"*/);
+import("c-bar-mobile"/* "KEY": "1"*/);
+```
+Invalid
+```js
+import("c-bar-mobile"/* "KEY SUBKEY": "SMALL"*/);
+import("c-bar-mobile"/* "😀": "1"*/);
+import("c-bar-mobile"/* @salesforce/client/formFactor = SMALL*/);
+
+import("c-bar-mobile" /* "@salesforce/client/formFactor": "SMALL && LARGE"*/);
+import("c-bar-mobile" /* "@salesforce/client/formFactor": "SMALL" && "@salesforce/client/formFactor": "LARGE"*/);
+
+// Dynamic module name specifier
+const bar = 'module-name';
+import(bar/* "@salesforce/client/formFactor": "SMALL" */)
+```
+
+#### Separation of Concern
+Dynamic imports hint metadata gathering happens during component compilation. An invariant of this process is that the metadata collection will **not** fail due to a bad hint. This could be due to a hint not following specified grammar or not using a well understood hint value.
+
+Additional validation and post processing of the gathered metadata is out of the scope of this RFC. That logic will be the responsibility of the application framework that consumes the metadata.
+
+Similarly, any design time validations of the hint grammar should be handled by linting code. For example, on salesforce platform, allowed hint keys might be restricted to only `@salesforce` scoped specifiers. The hint values will be validated based on the relevant value for the specifier. An ESLint rule will enforce this restriction via the [eslint-plugin-lwc](https://github.com/salesforce/eslint-plugin-lwc).
+
+### Reasons
+The hint syntax in the form of a comment was selected for the following reasons:
+* Complies with ‘what you write is what you see’ concept - an originally authored code remains untouched
+* Eliminates any additional code transformations at compile time
+* Eliminates any additional processing during module resolution at runtime (to strip out or pre-process module name)
+
+### Dynamic Import Metadata
+The proposed metadata format will include dynamically imported component name and a hint object, which in-itself contains the original query and a hint value. A hint value will be an array of objects with raw hint query, expected hint condition, and hint resource information.
+
+#### Proposed Shape
+```ts
+export interface SourceLocation {
+    startLine: number;
+    startColumn: number;
+    endline: number;
+    endColumn: number;
+    file: string;
+}
+
+export interface DynamicImportHint {
+    rawValue: string; // Trimmed hint phrase
+    hintKey: string;
+    hintValue: string;
+    location: SourceLocation;
+}
+
+export interface DynamicImportMetadata {
+    moduleName: string;
+    file: string;
```

Comment:
Why do we need a back pointer to the file here? The consumer of the `DynamicImportMetadata` should already be grouped by file.

Reply:
Metadata is gathered per BundleConfig. A BundleConfig can have more than one file. Currently, the metadata is not grouped by file. The relevant code is in getMetadata() routine of collector.ts
I will keep this in mind for our metadata factor project.